int32 actionId
int32 endCode
